### PR TITLE
Pr#1

### PR DESCRIPTION
**Pull request by Us10n**

### CODE REVIEW

File: repository/src/test/java/com/epam/esm/repository/dao/GiftCertificateDaoImplTest.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+//import org.springframework.test.context.junit.jupiter.SpringExtension;
+//
+//import java.time.LocalDateTime;
+//
+//@ExtendWith(SpringExtension.class)
+//@ContextConfiguration(classes = TestConfig.class)
+//@ActiveProfiles("test")
+//@Sql(value = "/createDb.sql", executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
+//@Sql(value = "/dropDb.sql", executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
+//class GiftCertificateDaoImplTest {
```
Comment (ZhukovAnton) [inline repository/src/test/java/com/epam/esm/repository/dao/GiftCertificateDaoImplTest.java:21]: Why this test commented?

File: service/src/main/java/com/epam/esm/service/converter/impl/GiftCertificateConverter.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import com.epam.esm.service.converter.DtoEntityConverter;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Component
+public class GiftCertificateConverter implements DtoEntityConverter<GiftCertificateDto, GiftCertificate> {
```
Comment (ZhukovAnton) [inline service/src/main/java/com/epam/esm/service/converter/impl/GiftCertificateConverter.java:19]: Please read about `mapstruct` lib. It's good solution for simple convertations from entity to dto and vise verse

**Review by ZhukovAnton — APPROVED** [review]
Note: the comments on this PR is not a trigger for fixes straight before the demo, but they are remarks for future

File: repository/src/main/java/com/epam/esm/repository/creator/GiftCertificateQueryCreator.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        String sortBy = criteria.getSortBy();
+
+        if (sortOrder != null && sortBy != null && sortBy.equalsIgnoreCase(CREATE_DATE)) {
+            if (sortOrder.equalsIgnoreCase(ASC)) {
+                criteriaQuery.orderBy(criteriaBuilder.asc(root.get(CREATE_DATE)));
+            }
+            if (sortOrder.equalsIgnoreCase(DESC)) {
+                criteriaQuery.orderBy(criteriaBuilder.desc(root.get(CREATE_DATE)));
+            }
+        }
```
Comment (HannaVatlina) [inline repository/src/main/java/com/epam/esm/repository/creator/GiftCertificateQueryCreator.java]: 1. Duplicate code
2. How can it be rewritten to make it more readable?

File: repository/src/main/java/com/epam/esm/repository/dao/impl/TagDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    private static final String FIND_WIDELY_USED_TAG_OF_USER_WITH_HIGHEST_COST_OF_ALL_ORDERS="SELECT tags.id, tags.name, gfs.name FROM tags " +
+            "JOIN gift_certificate_tags gfts on gfts.tag_id= tags.id " +
+            "JOIN gift_certificates gfs on gfs.id=gfts.gift_certificate_id " +
+            "JOIN orders ords on ords.gift_certificate_id=gfs.id " +
+            "JOIN users usrs on ords.user_id=usrs.id " +
+            "WHERE usrs.id=( " +
+            "    SELECT users.id FROM users JOIN orders ON orders.user_id=users.id GROUP BY users.id ORDER BY sum(orders.price) DESC LIMIT 1 " +
+            ") " +
+            "GROUP BY tags.name " +
+            "ORDER BY count(tags.name) DESC LIMIT 1";
```
Comment (HannaVatlina) [inline repository/src/main/java/com/epam/esm/repository/dao/impl/TagDaoImpl.java]: What if two or more tags have the same number of repetitions. How to show all of them?

File: service/src/main/java/com/epam/esm/service/service/impl/GiftCertificateServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+                .distinct()
+                .collect(Collectors.toList());
+        List<Tag> tagsToPersist = new ArrayList<>();
+        uniqueTags.forEach(newTag -> {
+            Optional<Tag> foundTag = tagDao.findByName(newTag.getName());
+            if (foundTag.isPresent()) {
+                tagsToPersist.add(foundTag.get());
+            } else {
+                tagsToPersist.add(newTag);
+            }
```
Comment (HannaVatlina) [inline service/src/main/java/com/epam/esm/service/service/impl/GiftCertificateServiceImpl.java]: Can replace it with `tagsToPersist.add(foundTag.orElse(newTag))`

File: service/src/main/java/com/epam/esm/service/service/impl/TagServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    @Override
+    public TagDto readById(long id) {
+        Optional<Tag> optionalTag = tagDao.findById(id);
+        if (!optionalTag.isPresent()) {
+            throw new NoSuchElementException(TAG_NOT_FOUND);
+        }
+
+        return tagConverter.convertToDto(optionalTag.get());
```
Comment (HannaVatlina) [inline service/src/main/java/com/epam/esm/service/service/impl/TagServiceImpl.java]: Think how can we replace it?

**Pull request merged**